updater/aws_test.go

```diff
@@ -15,6 +15,260 @@ import (
+		"inst-id-4": "{\"update_state\": \"Staged\", \"active_partition\": { \"image\": { \"version\": \"v1.1.1\"}}}",
+		"inst-id-5": "{\"update_state\": \"Available\", \"active_partition\": { \"image\": { \"version\": \"v1.0.5\"}}}",
+	}
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return &ssm.GetCommandInvocationOutput{
+				Status:                aws.String("Success"),
+				StandardOutputContent: aws.String(responses[*input.InstanceId]),
+			}, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return &ssm.SendCommandOutput{
+				Command: &ssm.Command{
+					CommandId:    aws.String("command-id"),
+					DocumentName: aws.String("check-document"),
+				},
+			}, nil
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM, checkDocument: "check-document"}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.NoError(t, err)
+	assert.Equal(t, expected, actual, "Should only contain instances in Aavailable or Ready update state")
+}
+
+func TestPaginatedFilterAvailableUpdatesSuccess(t *testing.T) {
+	checkPattern := "{\"update_state\": \"%s\", \"active_partition\": { \"image\": { \"version\": \"%s\"}}}"
+	expected := make([]instance, 0)
+	instances := make([]instance, 0)
+	getOut := &ssm.GetCommandInvocationOutput{
+		Status:                aws.String("Success"),
+		StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateAvailable, "v1.0.5")),
+	}
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+		expected = append(expected, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+			bottlerocketVersion: "v1.0.5",
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return getOut, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return &ssm.SendCommandOutput{
+				Command: &ssm.Command{
+					CommandId:    aws.String("command-id"),
+					DocumentName: aws.String("check-document"),
+				},
+			}, nil
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.NoError(t, err)
+	assert.EqualValues(t, expected, actual, "should contain all instances")
```

Comment:
I'd like to see a bit more assertion on what's actually happening within `filterAvailableUpdates`.

```suggestion
	sendCommandCalls := 0
	commandWaiterCalls := 0
	getCommandInvocationCalls := 0
	mockSSM := MockSSM{
		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
			getCommandInvocationCalls++
			return getOut, nil
		},
		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
			sendCommandCalls++
			return &ssm.SendCommandOutput{
				Command: &ssm.Command{
					CommandId:    aws.String("command-id"),
					DocumentName: aws.String("check-document"),
				},
			}, nil
		},
		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
			commandWaiterCalls++
			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
			return nil
		},
	}
	u := updater{ssm: mockSSM}
	actual, err := u.filterAvailableUpdates(instances)
	require.NoError(t, err)
	assert.EqualValues(t, expected, actual, "should contain all instances")
	assert.Equal(t, 2, sendCommandCalls, "should send commands for each page")
	assert.Equal(t, 100, commandWaiterCalls, "should wait for each instance")
	assert.Equal(t, 100, getCommandInvocationCalls, "should collect output for each instance")
```